File: docker/shared/path_guard.py
import os
from pathlib import Path
from typing import Literal

WRITE_ROOTS = ['sites', 'docs', 'ventures.json']


class PathGuard:
    def __init__(self, write_roots: list[str] | None = None, block_traversal: bool = True):
        self.write_roots = write_roots or WRITE_ROOTS
        self.block_traversal = block_traversal

    def _normalize(self, path: str) -> str:
        return Path(path).as_posix()

    def _has_traversal(self, path: str) -> bool:
        if not self.block_traversal:
            return False
        normalized = self._normalize(path)
        if normalized.startswith('/') or ':' in normalized:
            return True
        if normalized.startswith('~'):
            return True
        if '..' in normalized:
            return True
        if '/.' in normalized or normalized.startswith('.'):
            return True
        return False

    def _in_roots(self, path: str, roots: list[str]) -> bool:
        normalized = self._normalize(path)
        for root in roots:
            normalized_root = self._normalize(root)
            if normalized == normalized_root or normalized.startswith(normalized_root + '/'):
                return True
        return False

    def is_write_allowed(self, path: str) -> bool:
        if self._has_traversal(path):
            return False
        return self._in_roots(path, self.write_roots)

    def is_read_allowed(self, path: str) -> bool:
        if self._has_traversal(path):
            return False
        normalized = self._normalize(path)
        blocked = ['.env', '.env.local', '.env.production', '.git']
        for b in blocked:
            if normalized == b or normalized.startswith(b + '/'):
                return False
        if normalized.startswith('docker/secrets/'):
            return False
        return True

    def validate(self, operation: Literal['read', 'write'], path: str) -> dict:
        normalized = self._normalize(path)
        if self._has_traversal(path):
            return {'allowed': False, 'reason': f'Path traversal detected: {normalized}'}
        if operation == 'write':
            if not self.is_write_allowed(path):
                return {'allowed': False, 'reason': f'Write not allowed to: {normalized}. Allowed roots: {", ".join(self.write_roots)}'}
        else:
            if not self.is_read_allowed(path):
                return {'allowed': False, 'reason': f'Read not allowed to: {normalized}'}
        return {'allowed': True}

File: src/guards/path-guard.ts
import { resolve, normalize, relative } from 'path';

export interface PathGuardConfig {
  writeRoots: string[];
  readRoots: string[];
  blockTraversal: boolean;
}

export const WRITE_ROOTS = [
  'sites',
  'docs',
  'ventures.json',
];

export class PathGuard {
  private config: PathGuardConfig;

  constructor(config: PathGuardConfig) {
    this.config = config;
  }

  private normalizePath(p: string): string {
    return normalize(p).replace(/\\/g, '/');
  }

  private hasTraversal(p: string): boolean {
    if (!this.config.blockTraversal) return false;
    const normalized = this.normalizePath(p);
    if (normalized.startsWith('/') || /^[A-Z]:/i.test(normalized)) return true;
    if (normalized.startsWith('~')) return true;
    if (normalized.includes('..')) return true;
    if (normalized.includes('./') || normalized.includes('/.')) return true;
    return false;
  }

  private isInRoots(p: string, roots: string[]): boolean {
    const normalized = this.normalizePath(p);
    return roots.some(root => {
      const normalizedRoot = this.normalizePath(root);
      return normalized === normalizedRoot || normalized.startsWith(normalizedRoot + '/');
    });
  }

  isWriteAllowed(p: string): boolean {
    if (this.hasTraversal(p)) return false;
    return this.isInRoots(p, this.config.writeRoots);
  }

  isReadAllowed(p: string): boolean {
    if (this.hasTraversal(p)) return false;
    const blocked = ['.env', '.env.local', '.env.production', '.git'];
    const normalized = this.normalizePath(p);
    if (blocked.some(b => normalized === b || normalized.startsWith(b + '/'))) return false;
    if (normalized.startsWith('docker/secrets/')) return false;
    return this.isInRoots(p, this.config.readRoots);
  }

  validate(operation: 'read' | 'write', p: string): { allowed: boolean; reason?: string } {
    const normalized = this.normalizePath(p);

    if (this.hasTraversal(p)) {
      return { allowed: false, reason: `Path traversal detected: ${normalized}` };
    }

    if (operation === 'write') {
      if (!this.isWriteAllowed(p)) {
        return { allowed: false, reason: `Write not allowed to: ${normalized}. Allowed roots: ${this.config.writeRoots.join(', ')}` };
      }
    } else {
      if (!this.isReadAllowed(p)) {
        return { allowed: false, reason: `Read not allowed to: ${normalized}` };
      }
    }

    return { allowed: true };
  }
}
